prov/gni/src/gnix_buddy_allocator.c

```diff
@@ -0,0 +1,353 @@
+					   alloc_handle->base,
+					   MIN_BLOCK_SIZE,
+					   OFFSET(MIN_BLOCK_SIZE, j)));
+
+		dlist_insert_tail(tmp + OFFSET(MIN_BLOCK_SIZE, j),
+				  alloc_handle->lists + j);
+	}
+
+	/* Insert last block into list "i" */
+	dlist_insert_head(tmp, alloc_handle->lists + j);
+
+	/* Insert the buddy block of tmp into list "i" */
+	dlist_insert_tail(tmp + OFFSET(MIN_BLOCK_SIZE, j),
+			  alloc_handle->lists + j);
+}
+
+/**
+ * Find the first free block that can be split, then split it.
+ *
+ * @return 1  if the block cannot be found.
+ *
+ * @return 0 if the block is found.
+ */
+static inline int __gnix_buddy_find_block(handle_t *alloc_handle, size_t i)
+{
+	size_t j;
+
+	for (j = i + 1; j < alloc_handle->nlists; j++) {
+		if (!dlist_empty(alloc_handle->lists + j)) {
+			__gnix_buddy_split(alloc_handle, j, i);
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+
+/**
+ * If the buddy block is on the free list then coalesce and insert into the next
+ * list until we reach an allocated or split buddy block, or the max list size.
+ */
+static inline void __gnix_buddy_coalesce(handle_t *alloc_handle, void **ptr,
+					 size_t *block_size)
+{
+	while (*block_size < alloc_handle->max &&
+	       !_gnix_test_bit(&alloc_handle->bitmap,
+			       BITMAP_INDEX(BUDDY(*ptr, *block_size,
+						  alloc_handle->base),
+					    alloc_handle->base,
+					    MIN_BLOCK_SIZE,
+					    *block_size))) {
+
+		dlist_remove(BUDDY(*ptr, *block_size, alloc_handle->base));
+
+		/* Ensure ptr is the beginning of the new block */
+		if (*ptr > BUDDY(*ptr, *block_size, alloc_handle->base)) {
+			*ptr = BUDDY(*ptr, *block_size, alloc_handle->base);
+		}
+
+		*block_size *= 2;
+
+		_gnix_clear_bit(&alloc_handle->bitmap,
+				BITMAP_INDEX(*ptr, alloc_handle->base,
+					     MIN_BLOCK_SIZE, *block_size));
+	}
+
+}
+
+int _gnix_buddy_allocator_create(void *base, size_t len, size_t max,
+				 handle_t **alloc_handle)
+{
+	char err_buf[256] = {0}, *error = NULL;
+	int fi_errno;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	/* Ensure parameters are valid */
+	if (!base || !len || !max || max > len || !alloc_handle ||
+	    IS_NOT_POW_TWO(max) || (len % max) || !(len / MIN_BLOCK_SIZE * 2)) {
+
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid parameter to _gnix_buddy_allocator_create."
+			  "\n");
+		return -FI_EINVAL;
+	}
+
+	*alloc_handle = calloc(1, sizeof(handle_t));
+
+	if (!alloc_handle) {
+		error = strerror_r(errno, err_buf, sizeof(err_buf));
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Could not create buddy allocator handle.",
+			  error);
+		return -FI_ENOMEM;
+	}
+
+	fastlock_init(&alloc_handle[0]->lock);
+	fastlock_acquire(&alloc_handle[0]->lock);
+
+	alloc_handle[0]->base = base;
+	alloc_handle[0]->len = len;
+	alloc_handle[0]->max = max;
+
+	if (__gnix_buddy_create_lists(alloc_handle[0])) {
+		fastlock_release(&alloc_handle[0]->lock);
+		free(*alloc_handle);
+		return -FI_ENOMEM;
+	}
+
+	if ((fi_errno = _gnix_alloc_bitmap(&alloc_handle[0]->bitmap,
+					   len / MIN_BLOCK_SIZE * 2))) {
```

Comment:
This should be 2 * (Len / MBS).

Len is the number of bytes that the base block spans.

To have enough bits to flag each block as allocated or split whether the block size is 16B, 32B, 64B, ... we need:

Len / 16 bits for all the 16B blocks
Len / 32 bits for all the 32B blocks
Len / 64 bits for all the 64B blocks
...
Len / MaxBlockSize bits

So, as a maximum (if the max block size = Len) we need one less than two times Len / 16.
> On Jan 7, 2016, at 5:51 PM, James Swaro <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_buddy_allocator.c:
> 
> > +
> > +	fastlock_init(&alloc_handle[0]->lock);
> > +	fastlock_acquire(&alloc_handle[0]->lock);
> > +
> > +	alloc_handle[0]->base = base;
> > +	alloc_handle[0]->len = len;
> > +	alloc_handle[0]->max = max;
> > +
> > +	if (__gnix_buddy_create_lists(alloc_handle[0])) {
> > +		fastlock_release(&alloc_handle[0]->lock);
> > +		free(*alloc_handle);
> > +		return -FI_ENOMEM;
> > +	}
> > +
> > +	if ((fi_errno = _gnix_alloc_bitmap(&alloc_handle[0]->bitmap,
> > +					   len / MIN_BLOCK_SIZE * 2))) {
> why len / MBS *2?
> 
> —
> Reply to this email directly or view it on GitHub.
>